### PR TITLE
Oc session error channel monitoring enhancement for intrusive test

### PR DESCRIPTION
**Pull request by jun-rh**

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -84,8 +84,15 @@ func getOcSession(pod, container, namespace string, timeout time.Duration, optio
...
 		go func() {
-			err := <-outCh
-			gomega.Expect(err).To(gomega.BeNil())
+			for {
+				select {
+				case err := <-outCh:
+					log.Errorf("SpawnOc error: %v from %s %s", err, oc.GetPodName(), oc.GetPodContainerName())
+					gomega.Expect(err).To(gomega.BeNil())
+				case <-oc.GetDoneChannel():
+					break
```
Comment (edcdavid) [inline pkg/config/config.go:93]: Shouldn't we kill the cmd process here as well ?

**Review by edcdavid — APPROVED** [review]

File: pkg/config/config.go (hunk 1 above)
Comment (jun-rh) [reply pkg/config/config.go:93]: We could if we really want to release resources a little bit sooner. Probably not for now.

File: pkg/config/config.go [hunk 2]
```diff
@@ -84,8 +84,15 @@ func getOcSession(pod, container, namespace string, timeout time.Duration, optio
 		gomega.Expect(err).To(gomega.BeNil())
 		// Set up a go routine which reads from the error channel
 		go func() {
-			err := <-outCh
-			gomega.Expect(err).To(gomega.BeNil())
+			for {
+				select {
+				case err := <-outCh:
+					log.Errorf("SpawnOc error: %v from %s %s", err, oc.GetPodName(), oc.GetPodContainerName())
+					gomega.Expect(err).To(gomega.BeNil())
```
Comment (greyerof) [inline pkg/config/config.go:91]: Needs a previous defer ginkgo.GinkgoRecover() to avoid panic'ing in case err is not nil.

Comment (jun-rh) [reply pkg/config/config.go:91]: Hmm, I was wondering whether the original assertion is really the appropriate action here. Recover would let the test run continue, right? I don't think that's right either. Is there an alternative to panic if we just want to stop the test run?

Comment (hamadise) [reply pkg/config/config.go:91]: gingko.Fail() will only fail the test and continue, unless you have fast fail, in which case, the test would stop on the first failure.

File: test-network-function/lifecycle/suite.go [hunk 3]
```diff
@@ -275,54 +275,54 @@ func shutdownTest(podNamespace, podName string) {
...
+		ginkgo.By(fmt.Sprintf("test deployment in namespace %s", env.NameSpaceUnderTest))
+		deployments, notReadyDeployments = getDeployments(env.NameSpaceUnderTest)
+		if len(deployments) == 0 {
+			return
 		}
+		// We require that all deployments have the desired number of replicas and are all up to date
+		if len(notReadyDeployments) != 0 {
+			ginkgo.Skip("Can not test when deployments are not ready")
+		}
+		gomega.Expect(notReadyDeployments).To(gomega.BeEmpty())
```
Comment (hamadise) [inline test-network-function/lifecycle/suite.go:293]: line 293 code is not useful.. based on line 291, this test will always pass.

Comment (hamadise) [reply test-network-function/lifecycle/suite.go:293]: it can be removed safely

File: pkg/tnf/interactive/oc.go [hunk 4]
```diff
@@ -109,3 +111,13 @@ func (o *Oc) GetOptions() []Option {
 func (o *Oc) GetErrorChannel() <-chan error {
 	return o.errorChannel
 }
+
+// GetDoneChannel returns the receive only done channel
+func (o *Oc) GetDoneChannel() <-chan bool {
```
Comment (hamadise) [inline pkg/tnf/interactive/oc.go:116]: This function is not needed.. 
The moment you receive something in the channel, you can return in getOcSession function (line 92)
```
case <- oc :
break
```

Comment (jun-rh) [reply pkg/tnf/interactive/oc.go:116]: huh? oc is not a channel.

File: pkg/config/config.go (hunk 2 above)
Comment (jun-rh) [reply pkg/config/config.go:91]: Panic is the only way then? Personally I don't mind. It just looks a bit scary to people who don't know what it means. I could improve the log text a bit so they understand that the panic is intentional?

File: pkg/tnf/interactive/oc.go (hunk 4 above)
Comment (hamadise) [reply pkg/tnf/interactive/oc.go:116]: my bad. I thought you're getting the status of the channel and not the channel itself. And I was suggesting to get test directly against the channel. anyway, ignore this comment.

File: pkg/config/config.go (hunk 2 above)
Comment (hamadise) [reply pkg/config/config.go:91]: I'm not sure what are you trying to achieve here, but here's what to expect :
 * a failure in gomega or ginkgo.Fail would result in current test (ginkgo.It ) be stopped and the code will move to next test.
 * panic will stop everything.

Comment (jun-rh) [reply pkg/config/config.go:91]: I think we should stop the test altogether, as the test result wouldn't be useful if we continue running with broken sessions.

Comment (greyerof) [reply pkg/config/config.go:91]: Problem is this code is not part of any test: it's run inside the autodiscovery mechanism in the BeforeEach section. So, if the oc shell cannot be created, even using the defer ginkgo.GinkgoRecover() will make the first TC in the test suite and those after any intrusive TC to fail.
I'd go with removing the gomega assertion and use a logrus.Warn/Error message. These ocSessions are only used for the connectivity TCs, right?

Comment (jun-rh) [reply pkg/config/config.go:91]: No, they are used by other tests too.

**Review by hamadise — APPROVED** [review]

**Review by greyerof — APPROVED** [review]

**Pull request merged**